// luma.gl/dist/es5/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isBrowser", {
  enumerable: true,
  get: function get() {
    return _isBrowser.default;
  }
});
Object.defineProperty(exports, "isOldIE", {
  enumerable: true,
  get: function get() {
    return _isOldIe.default;
  }
});
Object.defineProperty(exports, "promisify", {
  enumerable: true,
  get: function get() {
    return _promiseUtils.promisify;
  }
});
Object.defineProperty(exports, "uid", {
  enumerable: true,
  get: function get() {
    return _utils.uid;
  }
});
Object.defineProperty(exports, "isPowerOfTwo", {
  enumerable: true,
  get: function get() {
    return _utils.isPowerOfTwo;
  }
});
Object.defineProperty(exports, "isObjectEmpty", {
  enumerable: true,
  get: function get() {
    return _utils.isObjectEmpty;
  }
});
Object.defineProperty(exports, "log", {
  enumerable: true,
  get: function get() {
    return _log.default;
  }
});
Object.defineProperty(exports, "formatValue", {
  enumerable: true,
  get: function get() {
    return _formatValue.formatValue;
  }
});
Object.defineProperty(exports, "getGLTypeFromTypedArray", {
  enumerable: true,
  get: function get() {
    return _typedArrayUtils.getGLTypeFromTypedArray;
  }
});
Object.defineProperty(exports, "getTypedArrayFromGLType", {
  enumerable: true,
  get: function get() {
    return _typedArrayUtils.getTypedArrayFromGLType;
  }
});
Object.defineProperty(exports, "flipRows", {
  enumerable: true,
  get: function get() {
    return _typedArrayUtils.flipRows;
  }
});
Object.defineProperty(exports, "scalePixels", {
  enumerable: true,
  get: function get() {
    return _typedArrayUtils.scalePixels;
  }
});

var _isBrowser = _interopRequireDefault(require("./is-browser"));

var _isOldIe = _interopRequireDefault(require("./is-old-ie"));

var _promiseUtils = require("./promise-utils");

var _utils = require("./utils");

var _log = _interopRequireDefault(require("./log"));

var _formatValue = require("./format-value");

var _typedArrayUtils = require("./typed-array-utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
