// @deck.gl/core/dist/es5/core/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "COORDINATE_SYSTEM", {
  enumerable: true,
  get: function get() {
    return _constants.COORDINATE_SYSTEM;
  }
});
Object.defineProperty(exports, "Deck", {
  enumerable: true,
  get: function get() {
    return _deck.default;
  }
});
Object.defineProperty(exports, "LayerManager", {
  enumerable: true,
  get: function get() {
    return _layerManager.default;
  }
});
Object.defineProperty(exports, "AttributeManager", {
  enumerable: true,
  get: function get() {
    return _attributeManager.default;
  }
});
Object.defineProperty(exports, "Layer", {
  enumerable: true,
  get: function get() {
    return _layer.default;
  }
});
Object.defineProperty(exports, "CompositeLayer", {
  enumerable: true,
  get: function get() {
    return _compositeLayer.default;
  }
});
Object.defineProperty(exports, "Viewport", {
  enumerable: true,
  get: function get() {
    return _viewport.default;
  }
});
Object.defineProperty(exports, "WebMercatorViewport", {
  enumerable: true,
  get: function get() {
    return _webMercatorViewport.default;
  }
});
Object.defineProperty(exports, "PerspectiveViewport", {
  enumerable: true,
  get: function get() {
    return _perspectiveViewport.default;
  }
});
Object.defineProperty(exports, "OrthographicViewport", {
  enumerable: true,
  get: function get() {
    return _orthographicViewport.default;
  }
});
Object.defineProperty(exports, "project", {
  enumerable: true,
  get: function get() {
    return _project.default;
  }
});
Object.defineProperty(exports, "project64", {
  enumerable: true,
  get: function get() {
    return _project2.default;
  }
});
Object.defineProperty(exports, "lighting", {
  enumerable: true,
  get: function get() {
    return _lighting.default;
  }
});
Object.defineProperty(exports, "View", {
  enumerable: true,
  get: function get() {
    return _view.default;
  }
});
Object.defineProperty(exports, "MapView", {
  enumerable: true,
  get: function get() {
    return _mapView.default;
  }
});
Object.defineProperty(exports, "FirstPersonView", {
  enumerable: true,
  get: function get() {
    return _firstPersonView.default;
  }
});
Object.defineProperty(exports, "ThirdPersonView", {
  enumerable: true,
  get: function get() {
    return _thirdPersonView.default;
  }
});
Object.defineProperty(exports, "OrbitView", {
  enumerable: true,
  get: function get() {
    return _orbitView.default;
  }
});
Object.defineProperty(exports, "PerspectiveView", {
  enumerable: true,
  get: function get() {
    return _perspectiveView.default;
  }
});
Object.defineProperty(exports, "OrthographicView", {
  enumerable: true,
  get: function get() {
    return _orthographicView.default;
  }
});
Object.defineProperty(exports, "MapController", {
  enumerable: true,
  get: function get() {
    return _mapController.default;
  }
});
exports.experimental = void 0;

require("./lib/init");

require("./shaderlib");

var _constants = require("./lib/constants");

var _deck = _interopRequireDefault(require("./lib/deck"));

var _layerManager = _interopRequireDefault(require("./lib/layer-manager"));

var _attributeManager = _interopRequireDefault(require("./lib/attribute-manager"));

var _layer = _interopRequireDefault(require("./lib/layer"));

var _compositeLayer = _interopRequireDefault(require("./lib/composite-layer"));

var _viewport = _interopRequireDefault(require("./viewports/viewport"));

var _webMercatorViewport = _interopRequireDefault(require("./viewports/web-mercator-viewport"));

var _perspectiveViewport = _interopRequireDefault(require("./viewports/perspective-viewport"));

var _orthographicViewport = _interopRequireDefault(require("./viewports/orthographic-viewport"));

var _project = _interopRequireDefault(require("./shaderlib/project/project"));

var _project2 = _interopRequireDefault(require("./shaderlib/project64/project64"));

var _lighting = _interopRequireDefault(require("./shaderlib/lighting/lighting"));

var _view = _interopRequireDefault(require("./views/view"));

var _mapView = _interopRequireDefault(require("./views/map-view"));

var _firstPersonView = _interopRequireDefault(require("./views/first-person-view"));

var _thirdPersonView = _interopRequireDefault(require("./views/third-person-view"));

var _orbitView = _interopRequireDefault(require("./views/orbit-view"));

var _perspectiveView = _interopRequireDefault(require("./views/perspective-view"));

var _orthographicView = _interopRequireDefault(require("./views/orthographic-view"));

var _mapController = _interopRequireDefault(require("./controllers/map-controller"));

var _orbitController = _interopRequireDefault(require("./controllers/orbit-controller"));

var _firstPersonState = _interopRequireDefault(require("./controllers/first-person-state"));

var _orbitState = _interopRequireDefault(require("./controllers/orbit-state"));

var _mapState = _interopRequireDefault(require("./controllers/map-state"));

var _viewportControls = _interopRequireDefault(require("./controllers/viewport-controls"));

var _mapControls = _interopRequireDefault(require("./controllers/map-controls"));

var _firstPersonViewport = _interopRequireDefault(require("./viewports/first-person-viewport"));

var _thirdPersonViewport = _interopRequireDefault(require("./viewports/third-person-viewport"));

var _orbitViewport = _interopRequireDefault(require("./viewports/orbit-viewport"));

var _effectManager = _interopRequireDefault(require("./experimental/lib/effect-manager"));

var _effect = _interopRequireDefault(require("./experimental/lib/effect"));

var _transitionManager = _interopRequireWildcard(require("./lib/transition-manager"));

var _linearInterpolator = _interopRequireDefault(require("./transitions/linear-interpolator"));

var _viewportFlyToInterpolator = _interopRequireDefault(require("./transitions/viewport-fly-to-interpolator"));

var _transitionUtils = require("./transitions/transition-utils");

var _log = _interopRequireDefault(require("./utils/log"));

var _get = require("./utils/get");

var _count = require("./utils/count");

var _binSorter = _interopRequireDefault(require("./utils/bin-sorter"));

var _colorUtils = require("./utils/color-utils");

var _scaleUtils = require("./utils/scale-utils");

var _flatten = require("./utils/flatten");

var _fp = require("./utils/fp64");

var _reflectionEffect = _interopRequireDefault(require("./experimental/reflection-effect/reflection-effect"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2015 - 2017 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

/* eslint-disable max-len */
// Intialize globals, check version
// Import shaderlib to make sure shader modules are initialized
// Core Library
// Experimental Pure JS (non-React) bindings
// Viewports
// Shader modules
// EXPERIMENTAL EXPORTS
// Experimental Features (May change in minor version bumps, use at your own risk)
// Controllers
// Experimental Controllers
// Experimental Effects (non-React) bindings
// Eperimental Transitions
// INTERNAL EXPORTS
// Layer utilities
// TODO - just expose as layer methods instead?
var experimental = {
  OrbitController: _orbitController.default,
  ViewportControls: _viewportControls.default,
  MapControls: _mapControls.default,
  FirstPersonState: _firstPersonState.default,
  OrbitState: _orbitState.default,
  MapState: _mapState.default,
  FirstPersonViewport: _firstPersonViewport.default,
  ThirdPersonViewport: _thirdPersonViewport.default,
  OrbitViewport: _orbitViewport.default,
  EffectManager: _effectManager.default,
  Effect: _effect.default,
  // Transitions
  TRANSITION_EVENTS: _transitionManager.TRANSITION_EVENTS,
  LinearInterpolator: _linearInterpolator.default,
  ViewportFlyToInterpolator: _viewportFlyToInterpolator.default,
  // For react module
  TransitionManager: _transitionManager.default,
  extractViewState: _transitionUtils.extractViewState,
  // TODO make this an internal export to set it apart from experimental
  // export const internal
  // For layers
  BinSorter: _binSorter.default,
  linearScale: _scaleUtils.linearScale,
  getLinearScale: _scaleUtils.getLinearScale,
  quantizeScale: _scaleUtils.quantizeScale,
  getQuantizeScale: _scaleUtils.getQuantizeScale,
  clamp: _scaleUtils.clamp,
  defaultColorRange: _colorUtils.defaultColorRange,
  log: _log.default,
  get: _get.get,
  count: _count.count,
  flatten: _flatten.flatten,
  countVertices: _flatten.countVertices,
  flattenVertices: _flatten.flattenVertices,
  fillArray: _flatten.fillArray,
  enable64bitSupport: _fp.enable64bitSupport,
  fp64ify: _fp.fp64ify,
  fp64LowPart: _fp.fp64LowPart,
  ReflectionEffect: _reflectionEffect.default
};
exports.experimental = experimental;
